### PR TITLE
Refactor library

### PR DESCRIPTION
**Pull request by esarafianou**
This PR fixes bugs introduced in #2 and adds one more feature as option.

**Highlights**

The middleware now overrides `res.writeHeads` as the initial version did. This is needed, because the middleware must be re-triggered after the user has added some data in `req.session` in order for the new cookie to be set.

It no longer uses `auth0-magic`, because all exported functions are asynchronous whereas `res.writeHeads` [cannot work in an asynchronous mode](https://github.com/expressjs/cookie-session/issues/75). The `aes-256-gcm` mode of the node `crypto` library is now used.

The new feature introduced is the `autoRenew` option. If set to true, a new cookie will be set in each response with an updated expiration date (`Date.now() + timeout`). If set to false, the expiration date will be set only the first time data are added to `req.session`.

**Lowlights**
- Added integration tests to avoid bugs like those introduced in the previous PR.
-  Added coverage in tests. The current coverage is 96.2% of lines.

### CODE REVIEW

File: lib/cookie-sessions.js [hunk 1]
```diff
@@ -32,147 +29,141 @@ var exports = module.exports = function(settings) {
...
-exports.decrypt = function(secret, ciphertext, nonce, cb) {
-  magic.decrypt.sync(secret, ciphertext, nonce, (err, plaintext) => {
-    if (err) {
-      return cb(err)
-    }
-    cb(null, plaintext.toString('utf-8'))
-  });
+exports.decrypt = function(secret, ciphertext, iv, authTag) {
+  try {
+    var decipher = crypto.createDecipheriv('aes-256-gcm', secret , Buffer.from(iv, 'hex'));
```
Comment (gkwang) [inline lib/cookie-sessions.js]: nit: extra space before comma

File: lib/cookie-sessions.js [hunk 2]
```diff
@@ -32,147 +29,141 @@ var exports = module.exports = function(settings) {
...
-    var jsonData = JSON.parse(cookieData)
-
-    if (!exports.valid(jsonData.t, timeout)) {
-      return cb(new Error('Cookie expired'))
-    }
-    cb(null, jsonData.d);
-  });
-};
+  }
+  var splited = exports.split(str)
```
Comment (gkwang) [inline lib/cookie-sessions.js]: nit spelling. should be splitted.

File: test/cookie-sessions.test.js [hunk 3]
```diff
@@ -17,104 +20,100 @@ describe('cookie sessions tests', () => {
...
   })
 
-  describe('readSession', () => {
-    it('session with key node_session read ok', () => {
-      sandbox.stub(sessions, 'deserialize').yields(null, {username: 'foobar'});
-      sessions.readSession('node_session', secret, 12, {cookies: {node_session: 'lala'}} , (err, data) => {
-        assert.ok(!err)
-        assert.deepEqual({username: 'foobar'}, data)
-      })
+  /*
```
Comment (gkwang) [inline test/cookie-sessions.test.js]: why did we comment this out?  If it's not needed, we should just remove it.

**Review by gkwang — APPROVED** [review]
Some nit comments; looks good otherwise. Merge at will.

File: test/cookie-sessions.test.js (hunk 3 above)
Comment (esarafianou) [reply test/cookie-sessions.test.js]: I removing it. I forgot to do so. There's no readSession function now.

**Pull request merged**